django/core/cache/backends/redis.py

```diff
@@ -0,0 +1,239 @@
+                self._pools[index] = self._pool_class(**params)
+
+        return self._pools[index]
+
+    def get_client(self, key='', write=False):
+        pool = self._get_connection_pool(write)
+        return self._client(connection_pool=pool)
+
+    def get(self, key, default):
+        client = self.get_client(key)
+        value = client.get(key)
+        # If key does not exists, return None
+        if value is None:
+            return default
+        else:
+            return self._serializer.loads(value)
+
+    def add(self, key, value, timeout):
+        client = self.get_client(key)
+        return bool(client.set(key, self._serializer.dumps(value), ex=timeout, nx=True))
+
+    def set(self, key, value, timeout):
+        client = self.get_client(key, write=True)
+        client.set(key, self._serializer.dumps(value), ex=timeout)
+
+    def touch(self, key, timeout):
+        client = self.get_client(key, write=True)
+        if timeout is None:
+            return client.persist(key)
+        else:
+            return client.expire(key, timeout)
+
+    def delete(self, key):
+        client = self.get_client(key, write=True)
+        return bool(client.delete(key))
+
+    def clear(self):
+        client = self.get_client(None, write=True)
+        return client.flushall()
+
+    def get_many(self, keys):
+        client = self.get_client(None)
+        ret = client.mget(keys)
+        return {
+            k: self._serializer.loads(v) for k, v in zip(keys, ret) if v is not None
+        }
+
+    def set_many(self, data, timeout):
+        client = self.get_client(None)
+        serialized_data = {k: self._serializer.dumps(v) for k, v in data.items()}
+
+        client.mset(serialized_data)
+
+        for k, _ in data.items():
+            # Setting timeout for each as redis-py does not support timeout with mset
+            if timeout is None:
+                client.persist(k)
+            else:
+                client.expire(k, timeout)
+
+
+class RedisCache(BaseCache):
+    def __init__(self, server, params):
+        super().__init__(params)
+        if isinstance(server, str):
+            self._servers = re.split('[;,]', server)
+        else:
+            self._servers = server
+
+        self._options = params.get('OPTIONS') or {}
+        self._class = self._options.get('CLIENT', RedisCacheClient)
+
+        if isinstance(self._class, str):
+            self._class = import_string(self._class)
+
+    def get_backend_timeout(self, timeout=DEFAULT_TIMEOUT):
+        # Needs rethinking
+
+        if timeout == DEFAULT_TIMEOUT:
+            timeout = self.default_timeout
+        elif timeout is None:
+            # Redis-py accepts None as timeout for as persistant key
+            return timeout
+        elif timeout <= 0:
+            # Need to fix this
+            timeout = 1
+
+        return int(timeout)
+
+    @cached_property
+    def _cache(self):
+        return self._class(self._servers, self._options)
```

Comment:
I think that we should unpack `self._options` here and make them arguments of `RedisCacheClient.__init__()`.

```suggestion
        return self._class(self._servers, **self._options)
```

This is how we approach this for all of the memcached backends using client classes implemented in third-party packages.

Reply:
Alright. So the changes would be something like
```
class RedisCacheClient:
    def __init__(self, servers, USERNAME=None, PASSWORD=None, ...):
        ...
        username = USERNAME
        ...
```
Something like this? 
Or we could work on a different way to manage the connection pool / authentication etc maybe?

Reply:
Yup, something like that. I'd go for lowercase options values though.

`OPTIONS`, as [documented](https://docs.djangoproject.com/en/3.2/topics/cache/#cache-arguments), can take some uppercase keys that are targeted at the non-memcache backends - mainly to deal with culling. But then it says:

> Memcached backends pass the contents of OPTIONS as keyword arguments to the client constructors, allowing for more advanced control of client behavior. For example usage, see below.

So `OPTIONS` is essentially passed to the client constructor, in our case - `RedisCacheClient`. If we look at the examples of options passed to `pylibmc` or `pymemcache` below that, they are all lowercase.

(You can update this to `Memcache and Redis backends...` with your documentation changes, by the way.)

With the options you've added so far, we're probably looking at:

```python
class RedisCacheClient:
    def __init__(
        self,
        servers,
        username=None,
        password=None,
        pool_class=None,
        parser_class=None,
        pickle_protocol=None,
    ):
        self._servers = servers
        self._pools = [None] * len(servers)

        self._pool_class = pool_class or redis.ConnectionPool
        self._client = redis.Redis
        self._serializer = PickleSerializer(pickle_protocol)

        if parser_class is None:
            parser_class = redis.connection.PythonParser
        elif isinstance(parser_class, str):
            parser_class = import_string(parser_class)

        self._client_kwargs = {'parser_class': parser_class}
        if username is not None:
            self._client_kwargs['username'] = username
        if password is not None:
            self._client_kwargs['password'] = password

        ...
```

I still need to look more into how Redis connections work and review that though, so I'll likely request other changes in this area. But hopefully the above gives an idea of what I was aiming at.